Profiles (#276)
oasis-tcs/csaf: csaf_2.0/prose/csaf-v2-editor-draft.md

Informatiove tests provide insights in common mistakes and bad practices. They MAY fail at a valid CSAF document. It is up to the issuing party to decide whether this was an intended behavior and can be ignore or should be treated. These tests may include information about recommended usage. A program MUST handle a test failure as a information.

# 5 Profiles

CSAF documents do not have many required fields as they can be used for different purposes. To ensure a common understanding which fields are required in a use case and therefore a certain document category each subsection defines a profile. It also provides insights into the purpose of the profile. Each profile extends the generic profile **Generic CSAF** making additional fields from the standard mandatory. Any other optional field from the standard can also be added to a CSAF document which conforms with a profile without breaking conformance with the profile. One and only exempt is when the profile requires not to have a certain set of fields.

## 5.1 Profile 1: Generic CSAF

> This profile defines the default required fields for any CSAF document. Therefore, it is a "catch all" for CSAF documents that do not satisfies any other profile. Furthermore, it is the foundation all other profiles are build on.

Reviewer: ```suggestion
This profile defines the default required fields for any CSAF document. Therefore, it is a "catch all" for CSAF documents that do not satisfy any other profile. Furthermore, it is the foundation all other profiles are build on.
```

Author: Thanks for catching that.

Informatiove tests provide insights in common mistakes and bad practices. They MAY fail at a valid CSAF document. It is up to the issuing party to decide whether this was an intended behavior and can be ignore or should be treated. These tests may include information about recommended usage. A program MUST handle a test failure as a information.

# 5 Profiles

> CSAF documents do not have many required fields as they can be used for different purposes. To ensure a common understanding which fields are required in a use case and therefore a certain document category each subsection defines a profile. It also provides insights into the purpose of the profile. Each profile extends the generic profile **Generic CSAF** making additional fields from the standard mandatory. Any other optional field from the standard can also be added to a CSAF document which conforms with a profile without breaking conformance with the profile. One and only exempt is when the profile requires not to have a certain set of fields.

Author: I made a suggestion for that.

## 5.1 Profile 1: Generic CSAF

This profile defines the default required fields for any CSAF document. Therefore, it is a "catch all" for CSAF documents that do not satisfy any other profile. Furthermore, it is the foundation all other profiles are build on.

A CSAF document SHALL fulfill the following requirements to satisfy the profile "Generic CSAF":

* The following elements must exist and be valid:
  * `/document/category`
  * `/document/publisher/category`
  * `/document/publisher/name`
  * `/document/title`
  * `/document/tracking/current_release_date`
  * `/document/tracking/id`
  * `/document/tracking/initial_release_date`
  * `/document/tracking/revision_history[]/date`
  * `/document/tracking/revision_history[]/number`
  * `/document/tracking/revision_history[]/summary`
  * `/document/tracking/status`
  * `/document/tracking/version`

> * The value of `/document/category` SHALL NOT be equal to any value that is intended to only be used by another profile. To explicitly select the use of this profile the value `generic_csaf` MAY be used.

Author: I added that remark in the intro as is applies for all profiles.

> An issuing party might choose to set `/document/publisher/name` in front of a value that is intended to only be used by another profile to state that the CSAF document does not use the profile associated with this value. This should be done if the issuing party is able or unwilling to use the value `generic_csaf`, e.g. due to legal or cooperate identity reasons.

Author: Yes. That was the idea. I'm not sure whether that is still needed as the profile "Security Advisory" is now characterized as `security_advisory`.

> An issuing party might choose to set `/document/publisher/name` in front of a value that is intended to only be used by another profile to state that the CSAF document does not use the profile associated with this value. This should be done if the issuing party is able or unwilling to use the value `generic_csaf`, e.g. due to legal or cooperate identity reasons.

Author: I adopted the examples.

> > The value `Example Company Security Advisory` in `/document/category` uses the profile "Generic CSAF". This is important to prepare forward compatibility as later versions of CSAF might add new profiles. Therefore, the values which can be used for the profile "Generic CSAF" might change.

Author: Yes. Currently, some companies use the category with the value "<Vendor> Security Advisory". We should explain somewhere somehow that this is not equivalent to the profile `security_advisory`.

I'm open to suggestions here.

> > The value `Example Company Security Advisory` in `/document/category` uses the profile "Generic CSAF". This is important to prepare forward compatibility as later versions of CSAF might add new profiles. Therefore, the values which can be used for the profile "Generic CSAF" might change.

Author: I adopted the examples.

## 5.2 Profile 2: Security incident response

This profile SHOULD be used to provide a response to a security breach or incident. This MAY also be used to convey information about an incident that is unrelated to the issuing party's own products or infrastructure.

> Example Company might use a CSAF document satisfying this profile to respond to a security incident at ACME Inc. and the implications on its own products and infrastructure.

A CSAF document SHALL fulfill the following requirements to satisfy the profile "Security incident response":

* The following elements must exist and be valid:
  * all elements required by the profile "Generic CSAF".
  * `/document/notes` with at least one item which has a `category` of `description`, `details`, `general` or `summary`

>     > Without at least one note item with information about response to the event referred to this doesn't provide any useful information.

Reviewer 2: Hard to read without relevant information. I suggest to remove this sentence.
```suggestion
```

Author: I added the `Reasoning` to separate that from implementation guidance.

  * `/document/references`

>     > There should be a `reference` (link) to a document or website which provides more details about the incident.

Reviewer 2: Contradicts with `must` above:
```suggestion
    > There MUST be a `reference` (link) to a document or website which provides more details about the incident.
```

Author: I rephrased that (also see https://github.com/oasis-tcs/csaf/pull/276#discussion_r660143398).

* The value of `/document/category` SHALL be `security_incident_response`.

## 5.3 Profile 3: Informational Advisory

This profile SHOULD be used to provide information which are **not related to a vulnerability** but e.g. a misconfiguration.

A CSAF document SHALL fulfill the following requirements to satisfy the profile "Informational Advisory":

* The following elements must exist and be valid:
  * all elements required by the profile "Generic CSAF".
  * `/document/notes` with at least one item which has a `category` of `description`, `details`, `general` or `summary`

>     > Without at least one note item with information about the "issue" which is the topic of the advisory it is useless.

Reviewer 2: As above, I would suggest to remove this sentence:
```suggestion
```

Author: I added the `Reasoning` to separate that from implementation guidance.

  * `/document/references`

>     > There should be a `reference` (link) to a document or website which provides more details about the issue or its remediation (if possible). This could be a link to a hardening guide, a manual, best practices or any other helpful information.

Reviewer 2: As above, contradicts with `must`:
```suggestion
    > There MUST be a `reference` (link) to a document or website which provides more details about the issue or its remediation (if possible). This could be a link to a hardening guide, a manual, best practices or any other helpful information.
```

Author: We should not have formal language in an informative comment. Personally, I like the "MUST" part but I don't know how you want to test that automatically... So I would rather go with Should in an informative comment... It doesn't contradict the must from above as the presence of the value and its content is separated. 

@sthagen: Would do you think?

Author: I rephrased that.

* The value of `/document/category` SHALL be `informational_advisory`.
* The element `/vulnerabilities` MUST NOT exist. If there is any information that would reside in the element `/vulnerabilities` the CSAF document SHOULD use another profile, e.g. "Security Advisory".

> If the `/product_tree` exists, a user MUST assume that all products mentioned are affected.

Reviewer 2: ```suggestion
If the element `/product_tree` exists, a user MUST assume that all products mentioned are affected.
```

## 5.4 Profile 4: Security Advisory

> This profile SHOULD be used to provide information which are related to vulnerabilities and corresponding remediations.

Reviewer 2: ```suggestion
This profile SHOULD be used to provide information which is related to vulnerabilities and corresponding remediations.
```

A CSAF document SHALL fulfill the following requirements to satisfy the profile "Security Advisory":

* The following elements must exist and be valid:
  * all elements required by the profile "Generic CSAF".
  * `/product_tree` which lists all products referenced later on in the CSAF document regardless of their state.
  * `/vulnerabilities[]/notes`
    > Provides details about the vulnerability.
  * `/vulnerabilities[]/product_status`

>     > Lists for each product its status in regard to the vulnerability.

Reviewer 2: ```suggestion
    > Lists each product's status in regard to the vulnerability.
```

* The value of `/document/category` SHALL be `security_advisory`.

## 5.5 Profile 5: VEX

> This profile SHOULD be used to provide information of the "Vulnerability Exploitability eXchange". The main purpose of the VEX format is to state that and why a certain product is not affected. See [VEX] for details.

Reviewer 2: ```suggestion
This profile SHOULD be used to provide information of the "Vulnerability Exploitability eXchange". The main purpose of the VEX format is to state that and why a certain product is, or is not, affected by a vulnerability. See [VEX] for details.
```

A CSAF document SHALL fulfill the following requirements to satisfy the profile "VEX":

* The following elements must exist and be valid:
  * all elements required by the profile "Generic CSAF".
  * `/product_tree` which lists all products referenced later on in the CSAF document regardless of their state.
  * at least one of
    * `/vulnerabilities[]/product_status/fixed`
    * `/vulnerabilities[]/product_status/known_affected`
    * `/vulnerabilities[]/product_status/known_not_affected`
    * `/vulnerabilities[]/product_status/under_investigation`
  * at least one of
    * `/vulnerabilities[]/cve`
    * `/vulnerabilities[]/id`
  * `/vulnerabilities[]/notes`
    > Provides details about the vulnerability.
* For each item in

>   * `/vulnerabilities[]/product_status/known_not_affected` an impact statement SHALL exist in `/vulnerabilities[]/threats`. The `category` value for such a statement MUST be `impact` and the `details` field SHALL contain a a description why the vulnerability cannot be exploited.

Reviewer 2: The VEX profile is in parts more strict than a security advisory, e.g. we require the use of `/vulnerabilites[]/threats` here. 
Is this intentional?

Author: That is a requirement of the VEX subgroup as they require impact statements. Every VEX document is a valid Security Advisory - but not every Security Advisory is a VEX document...

>   * `/vulnerabilities[]/product_status/known_affected` additional product specific information SHALL be provided in `/vulnerabilities[]/remediations` as an action statement. Optional, additional information MAY also be provide through `/vulnerabilities[]/notes` and `/vulnerabilities[]/threats`.

Reviewer 2: This means that a vendor must already know of remediations when the document is published. I do not like this, as it would prevent early notifications. I suggest to relax this requirement to "SHOULD".

Author: You can always select `none_available` or `no_fix_planned`.

Author: We discussed that in the VEX group and they were happy with the SHALL so I don't really want to change that.

>   > Even though Product status lists Product IDs, Product Group IDs can be used in the `remediations` and `threats` object. However, it MUST be ensured that for each Product ID the required information is available.

Reviewer 2: What is the required information?

Author: Basically, the action or impact statement. I rephrased that. Is it clear now?

Reviewer 2: For me at least the term 'required information' is somehow misleading. How about: 
```
However, it MUST be ensured that the referenced vulnerability information is consistent and complete for each Product ID.
```

Author: Well, that is contentwise not the same as:

`However, it MUST be ensured that for each Product ID the required information as stated above is available according to its product status.`

1. Just from reading the sentence I'm not sure whether I have the correct understanding of `the referenced vulnerability information`. Is it `remediations` and `threats`?
2. `consistent and complete`: I get the idea. However, these words might be a little bit too big. How do I enforce that in a text field?

Unfortunately, I seem to fail here to make the (obviously) required information clearer. Maybe, I was too reluctant to directly name them since they are listed as the points right above the informative comment.

What about that version:

`However, it MUST be ensured that for each Product ID the required information according to its product status as stated in the two points above is available.`

If necessary, we could add the following:

`This implies that all products with the status `known_not_affected` MUST have an impact statement and all products with the status `known_affected` MUST have additional product specific information regardless whether that is referenced through the Product ID or a Product Group ID.`
